File: 05-Algorithms/code/typescript/src/bfs.ts
import { buildAdjList, checkVertexInRange, type Edge } from "./graphCommon.js";

/**
 * Breadth-first traversal from `src` over n vertices (0..n-1). Returns the
 * order vertices were first visited in, starting with `src`. Neighbors are
 * always explored in ascending vertex-id order (see buildAdjList), so the
 * result is fully deterministic. Unreached vertices are simply absent from
 * the result. O(V + E).
 */
export function bfs(n: number, edges: Edge[], src: number, directed = false): number[] {
  checkVertexInRange(n, src);
  const adj = buildAdjList(n, edges, directed);
  const visited = new Array(n).fill(false);
  const order: number[] = [];
  const queue: number[] = [src];
  visited[src] = true;
  let head = 0;
  while (head < queue.length) {
    const u = queue[head++];
    order.push(u);
    for (const v of adj[u]) {
      if (!visited[v]) {
        visited[v] = true;
        queue.push(v);
      }
    }
  }
  return order;
}

File: 05-Algorithms/code/python/seb_alg/bfs.py
"""Mirrors `typescript/src/bfs.ts` exactly."""

from __future__ import annotations

from typing import List, Sequence

from .graph_common import build_adj_list, check_vertex_in_range


def bfs(n: int, edges: Sequence[Sequence[int]], src: int, directed: bool = False) -> List[int]:
    """Breadth-first traversal from src. Returns visit order. O(V + E)."""
    check_vertex_in_range(n, src)
    adj = build_adj_list(n, edges, directed)
    visited = [False] * n
    order: List[int] = []
    queue: List[int] = [src]
    visited[src] = True
    head = 0
    while head < len(queue):
        u = queue[head]
        head += 1
        order.append(u)
        for v in adj[u]:
            if not visited[v]:
                visited[v] = True
                queue.append(v)
    return order
